### PR TITLE
Client grader submissions parsing

### PR DESCRIPTION
**Pull request by jubustamantem**
- Added new functions "from_dict" and "client_grader_result_to_html_block".
- Added a new mode "dict" in class "ParsableText". This should be the "response_type" for client grader submissions.
- To fix the "internal error" parsing, BSON decoding is avoided if "grader_results" is found in the submission while getting input.
- New condition was added in "get_feedback_from_submission" to use the new parsing functions.

### CODE REVIEW

File: inginious/frontend/parsable_text.py [hunk 1]
```diff
@@ -334,6 +350,69 @@ def hdl_to_html_block(self, test_id, result, test_case, input_sample, debug_info
...
+        for i in range(0, len(test["id"])):
+            if test["id"][i].isdigit():
+                id_num += " " + test["id"][i:]
+                break
+
+        template_info = {
+            "test_id": id_num,
+            "test_grade": test["test_grade"],
+            "test_message": test["test_message"],
+            "panel_id": "collapseDebug" + str(test["id"][-1]),
```
Comment (ferestrepoca) [inline inginious/frontend/parsable_text.py]: This part is not clear for me. Maybe:

```suggestion
            "panel_id": "collapseDebug" + id_num,
```

File: inginious/frontend/i18n/es/LC_MESSAGES/messages.po [hunk 2]
```diff
@@ -68,41 +68,78 @@ msgstr ""
...
 msgstr ""
 "<br><strong>{}:</strong> Hubo un error mientras se ejecutaba el notebook:"
 " <br><pre>{}</pre><br>"
 
-#: inginious/frontend/parsable_text.py:357
+#: inginious/frontend/parsable_text.py:372
+msgid ""
+"<ul class=\"list_disc\" style=\"font-size:12px;\"><li>\n"
+"            <strong style=\"font-size:15px\"> Test{test_id}: </strong><i>"
+" - YOUR GRADE = {test_grade} / 100 </i>"
```
Comment (ferestrepoca) [inline inginious/frontend/i18n/es/LC_MESSAGES/messages.po:95]:
```suggestion
" - Calificación = {test_grade} / 100 </i>"
```

File: inginious/frontend/i18n/es/LC_MESSAGES/messages.po [hunk 3]
```diff
@@ -68,41 +68,78 @@ msgstr ""
...
+msgstr ""
+
+#: inginious/frontend/parsable_text.py:377
+msgid ""
+"<a class=\"btn btn-default btn-link btn-xs\" role=\"button\"\n"
+"            data-toggle=\"collapse\" href=\"#{panel_id}\" aria-"
+"expanded=\"false\" aria-controls=\"{panel_id}\">\n"
+"            View Details </a><div class=\"collapse\" id=\"{panel_id}\"> "
```
Comment (ferestrepoca) [inline inginious/frontend/i18n/es/LC_MESSAGES/messages.po:106]:
```suggestion
"            Ver detalles </a><div class=\"collapse\" id=\"{panel_id}\"> "
```

File: inginious/frontend/i18n/es/LC_MESSAGES/messages.po [hunk 4]
```diff
@@ -68,41 +68,78 @@ msgstr ""
...
+"<pre>{test_message}</pre> <br>\n"
+"            <ul class=\"list_disc\" style=\"font-size:13px;\">"
+msgstr ""
+
+#: inginious/frontend/parsable_text.py:384
+msgid "<li><strong style=\"font-size:14px\"> FUNCTIONS </strong> <br>"
```
Comment (ferestrepoca) [inline inginious/frontend/i18n/es/LC_MESSAGES/messages.po:118]:
```suggestion
msgid "<li><strong style=\"font-size:14px\"> FUNCIONES </strong> <br>"
```

**Review by ferestrepoca — CHANGES_REQUESTED** [review]
Thanks, Juan. Just some small suggestions. Please check them out.

File: inginious/frontend/parsable_text.py [hunk 5]
```diff
@@ -334,6 +350,69 @@ def hdl_to_html_block(self, test_id, result, test_case, input_sample, debug_info
...
+        """Parse each test of client grader submissions into html"""
+
+        test_functions = json.loads(test["functions"].replace("'", "\""), strict=False)
+        test_variables = json.loads(test["variables"].replace("'", "\""), strict=False)
+
+        id_num = ""
+        for i in range(0, len(test["id"])):
+            if test["id"][i].isdigit():
+                id_num += " " + test["id"][i:]
+                break
```
Comment (ferestrepoca) [inline inginious/frontend/parsable_text.py]: What if `test["id"]`is only made of alphabetic characters? Is it possible? In that case the result in `id_num` will be `''`.

Comment (VirtualDiegox) [inline inginious/frontend/parsable_text.py]: This part is unnecessary, the test id could be any alphanumeric string, it is just an identifier for the test assigned by the teacher

File: inginious/frontend/parsable_text.py (hunk 1 above)
Comment (VirtualDiegox) [inline inginious/frontend/parsable_text.py]: I see that id_num logic is a string with a blank space plus take the first number on the test id string, so i'd go this way around removing id_num logic, panel_id it is just useful to dropdown work independently
```suggestion
        template_info = {
            "test_id": test["id"],
            "test_grade": test["test_grade"],
            "test_message": test["test_message"],
            "panel_id": "collapseDebug" + test["id"],
```

File: inginious/frontend/submission_manager.py [hunk 6]
```diff
@@ -383,8 +386,13 @@ def get_feedback_from_submission(self, submission, only_feedback=False, show_eve
         """
         if only_feedback:
             submission = {"text": submission.get("text", None), "problems": dict(submission.get("problems", {}))}
+        
         if "text" in submission:
             submission["text"] = ParsableText(submission["text"], submission.get("response_type", "rst"), show_everything, translation).parse()
+
+        if "grader_results" in submission:
+            submission["text"] = ParsableText(submission, submission.get("response_type"), show_everything, translation).parse()
```
Comment (VirtualDiegox) [inline inginious/frontend/submission_manager.py]: add this only to ensure the logic flow, if grader_results in submmions, the only response type acceptable is dict, and if response_type is not found on the submission the default parse is rst, so better ensure this.
```suggestion
            submission["text"] = ParsableText(submission, submission.get("response_type","dict"), show_everything, translation).parse()
```

File: inginious/frontend/parsable_text.py [hunk 7]
```diff
@@ -216,7 +218,21 @@ def html(cls, string, show_everything=False, translation=gettext.NullTranslation
         out, _ = tidylib.tidy_fragment(string)
         return out
     
-    
+    def from_dict(self, dictionary, show_everything=False, translation=gettext.NullTranslations()):
+        """Parses DICT"""
+        # Load object
+
+        grader_results = dictionary["grader_results"]
```
Comment (VirtualDiegox) [inline inginious/frontend/parsable_text.py]: Just in case could be:
```suggestion
        grader_results = dictionary.get("grader_results",{})
```

**Review by VirtualDiegox — CHANGES_REQUESTED** [review]
I just have a few suggestions, LGTM with the requested changes

File: inginious/frontend/parsable_text.py (hunk 5 above)
Comment (jubustamantem) [inline inginious/frontend/parsable_text.py]:
```suggestion
```
Remove unnecessary code

File: inginious/frontend/parsable_text.py (hunk 1 above)
Comment (jubustamantem) [reply inginious/frontend/parsable_text.py]: Solved with https://github.com/JuezUN/INGInious/pull/504#discussion_r1136141990

**Review by ferestrepoca — APPROVED** [review]
LGTM. Nice work. Thanks!

File: inginious/frontend/i18n/es/LC_MESSAGES/messages.po [hunk 8]
```diff
@@ -68,41 +68,87 @@ msgstr ""
...
+#, fuzzy
+msgid ""
+"<strong>: </strong><pre class=\"language-python\"><code\n"
+"                class=\"language-python\" data-"
+"language=\"python\"></code></pre><script>highlight_code();</script>"
+msgstr ""
+"<strong>Código ejecutado:</strong><pre class=\"language-python\"><code "
+"class=\"language-python\" data-"
+"language=\"python\">{case_code}</code></pre><script>highlight_code();</script>"
+
```
Comment (VirtualDiegox) [inline inginious/frontend/i18n/es/LC_MESSAGES/messages.po]: Careful here, normally translations with the fuzzy tag mean that pybabel is not sure about it, the best practice is check fuzzy tags and fix them, here for example in the translation, there is an extra parameter {case_code} on line 135 that is not on the msgid, remove it and check if it works without problem

**Review by VirtualDiegox — APPROVED** [review]

**Pull request merged**